iroh-net/src/magicsock.rs

```diff
@@ -2051,62 +1972,73 @@ impl Actor {
     /// never be invoked directly.  Some day this will be refactored to not allow this easy
     /// mistake to be made.
     #[instrument(level = "debug", skip_all)]
-    async fn update_direct_addrs(&mut self, why: &'static str) {
+    async fn refresh_direct_addrs(&mut self, why: &'static str) {
         inc!(MagicsockMetrics, update_direct_addrs);
 
         debug!("starting direct addr update ({})", why);
         self.port_mapper.procure_mapping();
         self.update_net_info(why).await;
     }
 
-    /// Stores the results of a successful direct addr update.
-    async fn store_direct_addr_update(&mut self, nr: Option<Arc<netcheck::Report>>) {
+    /// Updates the direct addresses of this magic socket.
+    ///
+    /// Updates the [`DiscoveredDirectAddrs`] of this [`MagicSock`] with the current set of
+    /// direct addresses from:
+    ///
+    /// - The portmapper.
+    /// - A netcheck report.
+    /// - The local interfaces IP addresses.
+    fn update_direct_addresses(&mut self, netcheck_report: Option<Arc<netcheck::Report>>) {
         let portmap_watcher = self.port_mapper.watch_external_address();
 
-        // direct_addr -> how it was found
-        let mut already = HashMap::new();
-        // unique direct addrs
-        let mut addrs = Vec::new();
-
-        macro_rules! add_addr {
-            ($already:expr, $addrs:expr, $ipp:expr, $typ:expr) => {
-                #[allow(clippy::map_entry)]
-                if !$already.contains_key(&$ipp) {
-                    $already.insert($ipp, $typ);
-                    $addrs.push(DirectAddr {
-                        addr: $ipp,
-                        typ: $typ,
-                    });
-                }
-            };
-        }
+        // We only want to have one DirectAddr for each SocketAddr we have.  So we store
+        // this as a map:
+        let mut addrs: BTreeMap<SocketAddr, DirectAddr> = BTreeMap::new();
```

Comment:
nitpick: since `DirectAddr` is just a `SocketAddr` and `DirectAddrType`, couldn't `addrs` be a `BTreeMap<SocketAddr, DirectAddrType>` & then you can map & collect it when you call `msock.store_direct_addresses`?

Reply:
this is an excellent suggestion!